### PR TITLE
Match AB frame

### PR DESCRIPTION
**Pull request by tiffanyhsyu**
Introduces a NIRSPEC settings file and adds an 'imagetype' keyword to help better sort the NIRSPEC calibration and science frames. For objects classified as 'science', finds a matching A/B frame and prints it in the output .pypit file.

### CODE REVIEW

File: pypit/arload.py [hunk 1]
```diff
@@ -105,12 +105,15 @@ def generate_updates(dct, keylst, keys, whddict, headarr):
             kchk = '.'.join(ch.split('.')[1:])
             frhd = whddict['{0:02d}'.format(tfrhd)]
             # JFH changed to in instead of !=
-            if ((settings_spect['check'][ch] in str(headarr[frhd][kchk]).strip()) == False):
-                print(ch, frhd, kchk)
-                print(settings_spect['check'][ch], str(headarr[frhd][kchk]).strip())
-                msgs.warn("The following file:"+msgs.newline()+datlines[i]+msgs.newline()+"is not taken with the settings.{0:s} detector".format(settings_argflag['run']['spectrograph'])+msgs.newline()+"Remove this file, or specify a different settings file.")
-                msgs.warn("Skipping the file..")
-                skip = True
+            try:
```
Comment (profxj) [inline pypit/arload.py]: can we remove the try/except now?

File: pypit/core/arsort.py [hunk 2]
```diff
@@ -528,6 +530,91 @@ def build_frametype_list(fitstbl):
     # Return
     return ftypes
 
+def match_ABBA(fitstbl):
```
Comment (profxj) [inline pypit/core/arsort.py]: Add a doc string for the method briefly describing it.

File: pypit/core/arsort.py [hunk 3]
```diff
@@ -528,6 +530,91 @@ def build_frametype_list(fitstbl):
...
+        # Turn mask 'off' (i.e., to True) now that they have been grouped to this target
+        mask[targets[targ_name]] = True
+        #print(mask)
+
+    # Create an empty list of filenames that will store each frame's buddy A/B frame
+    AB_frame = [''] * len(fitstbl)
+
+    for key, value in targets.items():
+
+        #print(key, value)
```
Comment (profxj) [inline pypit/core/arsort.py]: Remove this and other commented out print() statements.

File: pypit/core/arsort.py [hunk 4]
```diff
@@ -528,6 +530,91 @@ def build_frametype_list(fitstbl):
...
+        value_groups = [value[i:i + 4] for i in range(0, len(ABBA_coords), 4)]
+
+        for group in range(len(ABBA_groups)):
+            # Warn user that if there are any groups of only 2 files, assuming they are in order of A and B
+            if len(ABBA_groups[group]) == 2:
+                msgs.info('Assuming these two frames are A and B frame.')
+            # Check that we have a 4-file, ABBA sequence
+            elif len(ABBA_groups[group]) == 4:
+                # Check that frames 1, 4 of an ABBA sequence are at the same nod position (A) based on their RA, DEC
+                if not ABBA_coords[0].separation(ABBA_coords[-1]).deg < 5e-4:
```
Comment (profxj) [inline pypit/core/arsort.py]: Maybe make this a TOL parameter and have it performed
in arcsec instead of deg.

File: pypit/core/arsort.py [hunk 5]
```diff
@@ -528,6 +530,91 @@ def build_frametype_list(fitstbl):
...
+                    print(ABBA_groups[group])
```
Comment (profxj) [inline pypit/core/arsort.py]: use msgs.info() or remove

File: pypit/core/arsort.py (hunk 4 above)
Comment (profxj) [inline pypit/core/arsort.py]: You can avoid the not by using > instead of <

File: pypit/core/arsort.py [hunk 6]
```diff
@@ -528,6 +530,91 @@ def build_frametype_list(fitstbl):
...
+                    msgs.warn('Are frames ' + str((group * 4) + 1) + ' and ' + str(
+                        (group * 4) + 4) + ' for target ' + key + ' both A frames?')
+                # Check that frames 2, 3 of an ABBA sequence are both at the same nod position (B)
+                if not ABBA_coords[1].separation(ABBA_coords[2]).deg < 5e-4:
+                    msgs.warn('Are frames ' + str((group * 4) + 2) + ' and ' + str(
+                        (group * 4) + 3) + ' for target ' + key + ' both B frames?')
+            # Perhaps this isn't needed..? Should files either be in 4 or 2 at this point?
```
Comment (profxj) [inline pypit/core/arsort.py]: Agreed. If you get here, I think it is a bug and you might
as well raise a  msgs.error()

File: pypit/core/arsort.py [hunk 7]
```diff
@@ -528,6 +530,91 @@ def build_frametype_list(fitstbl):
...
+
+            # Create a copy of the array value_groups[group] (which gives the indices corresponding to the 4/2 ABBA/AB files in consideration)
+            AB_idx_flip = np.copy(value_groups[group])
+            # Flip such that, for example, (1, 2, 3, 4) --> (2, 1, 4, 3)
+            AB_idx_flip[::2], AB_idx_flip[1::2] = value_groups[group][1::2], value_groups[group][::2]
+
+            # Fill in AB_frame list
+            for i in range(len(value_groups[group])):
+                AB_frame[value_groups[group][i]] = fitstbl['filename'][AB_idx_flip[i]]
+
```
Comment (profxj) [inline pypit/core/arsort.py]: We might add to the fitstbl here, i.e.

fitstbl['AB_frame'] = AB_frame

File: pypit/pypitsetup.py [hunk 8]
```diff
@@ -4,7 +4,7 @@
 import inspect
 import numpy as np
 
-#from importlib import reload
+from importlib import reload
```
Comment (profxj) [inline pypit/pypitsetup.py]: comment out again

File: pypit/pypitsetup.py [hunk 9]
```diff
@@ -78,6 +78,7 @@ def build_fitstbl(self, file_list):
         fitstbl : Table
 
         """
+        reload(arload)
```
Comment (profxj) [inline pypit/pypitsetup.py]: remove

File: pypit/pypitsetup.py [hunk 10]
```diff
@@ -96,8 +97,10 @@ def build_group_dict(self):
           Dict describing the various setups
         """
         #
-        all_sci_idx = self.fitstbl['sci_ID'].data[self.fitstbl['science']]
-        self.group_dict = arsetup.build_group_dict(self.fitstbl, self.setupIDs, all_sci_idx)
+        reload(arsetup)
```
Comment (profxj) [inline pypit/pypitsetup.py]: remove reload()

**Comment by profxj** [comment]
Thanks for the tests @tiffanyhsyu !

I've added the NIRSPEC settings file to the archive 
which should get us passing Travis.  But I am failing the
test_pypitsetup.py test on my machine with this error:

    
        if not indices:
>           raise KeyError("Keyword {!r} not found.".format(keyword))
E           KeyError: "Keyword 'DSENSOR' not found."

Are you passing?

**Comment by tiffanyhsyu** [comment]
Ok, I'm not too Travis savvy, but I failed on 5 tests previously (4 hours ago) and just looked at them and they were all:

[WARNING] :: No archival files found for /home/travis/build/PYPIT/PYPIT/pypit/data/settings/settings.keck_nirspec

So I'm guessing I should be good now that you've archived the settings file. But how do I rerun a Travis test?

**Comment by profxj** [comment]
Travis is passing, it is an internal test on

**Comment by profxj** [comment]
merging.

**Pull request merged**